Add cicd docs (#313)
kyma-project/istio: docs/contributor/04-30-ci-cd.md

# CI/CD

> 

Reviewer: ```suggestion
The Istio module contains several pipelines that run on pull requests, on the main branch, on the release branch, and periodically. They perform various tasks such as running a linter, executing unit tests with code coverage, generating artifacts for development and production purposes, and running integration tests. The integration tests ensure the functional correctness of Istio Manager and verify if upgrading to the new version of Istio Manager is possible.

```

| [`Pull Request integration tests / Kubernetes version compatibility test`](https://github.com/kyma-project/istio/blob/main/.github/workflows/pull-integration.yaml#L33) | Executes basic integration test suite for Istio Module and checks if any incompatibilities occurred during the execution.                       |
| [`Pull Request integration tests / Istio upgrade integration test`](https://github.com/kyma-project/istio/blob/main/.github/workflows/pull-integration.yaml#L44)        | Executes upgrade integration test suite for Istio Module. It checks if upgrade of existing release to the new version can be done successfully. |
| [`Pull Request / lint`](https://github.com/kyma-project/istio/blob/main/.github/workflows/pull-request.yaml#L24)                                                        | Responsible for linting, static code analysis.                                                                                                  |

> | [`Pull Request / Run unit tests`](https://github.com/kyma-project/istio/blob/main/.github/workflows/pull-request.yaml#L39)                                              | Responsible for running unit tests with code coverage information.                                                                              |

Reviewer: ```suggestion
| [`Pull Request / Run unit tests`](https://github.com/kyma-project/istio/blob/main/.github/workflows/pull-request.yaml#L39)                                              | Runs unit tests that generate code coverage information.                                                                              |
```

## Pipelines running on main branch

> The following CI jobs run on main branch. They generate production ready artifacts and run integration tests against them to verify functional correctness of Istio Manager including possibility of upgrade to the new version.

Reviewer: ```suggestion
The following CI jobs run on the main branch. They generate production-ready artifacts and execute integration tests against them to verify the functional correctness of Istio Manager. These tests also check if upgrading to the new version of Istio Manager is possible.
```

| Name                                                                                                                                                      | Description                                                                                                                                 |
|-----------------------------------------------------------------------------------------------------------------------------------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------|

> | [`post-istio-operator-build`](https://github.com/kyma-project/test-infra/blob/main/prow/jobs/istio/istio-manager.yaml#L308)                               | Builds Istio operator image and pushes it into the prod registry.                                                                           |

Reviewer: ```suggestion
| [`post-istio-operator-build`](https://github.com/kyma-project/test-infra/blob/main/prow/jobs/istio/istio-manager.yaml#L308)                               | Builds Istio Operator's image and pushes it to the `prod` registry.                                                                           |
```

> | [`post-istio-module-build`](https://github.com/kyma-project/test-infra/blob/main/prow/jobs/istio/istio-manager.yaml#L163)                                 | Builds ModuleTemplate for the Istio module and pushes it to the prod registry.                                                              |

Reviewer: ```suggestion
| [`post-istio-module-build`](https://github.com/kyma-project/test-infra/blob/main/prow/jobs/istio/istio-manager.yaml#L163)                                 | Builds ModuleTemplate for the Istio module and pushes it to the `prod` registry.                                                              |
```

| [`post-istio-integration`](https://github.com/kyma-project/test-infra/blob/main/prow/jobs/istio/istio-integration.yaml#L87)                               | Executes main integration test suite for Istio Manager on k3d cluster.                                                                      |

> | [`main integration tests / Wait for image build`](https://github.com/kyma-project/istio/blob/main/.github/workflows/main-integration.yaml#L20)            | Waits for operator image build to finish so the image dependent jobs can be executed.                                                       |

Reviewer: ```suggestion
| [`main integration tests / Wait for image build`](https://github.com/kyma-project/istio/blob/main/.github/workflows/main-integration.yaml#L20)            | Waits for the operator's image build to finish so that image-dependent jobs can be executed.                                                       |
```

> | [`main integration tests / Istio upgrade integration tests`](https://github.com/kyma-project/istio/blob/main/.github/workflows/main-integration.yaml#L30) | Executes upgrade integration test suite of Istio Module. It checks if upgrade existing release to the new version can be done successfully. |

Reviewer: ```suggestion
| [`main integration tests / Istio upgrade integration tests`](https://github.com/kyma-project/istio/blob/main/.github/workflows/main-integration.yaml#L30) | Executes the upgrade integration test suite of Istio Module and verifies if the existing release can be successfully upgraded to the new version. |
```

> | [`main integration tests / Slack Notification`](https://github.com/kyma-project/istio/blob/main/.github/workflows/main-integration.yaml#L42)              | In case of failed pipelines on main branch it sends Slack notification to the team's channel.                                               |

Reviewer: ```suggestion
| [`main integration tests / Slack Notification`](https://github.com/kyma-project/istio/blob/main/.github/workflows/main-integration.yaml#L42)              | Sends a Slack notification to the team's channel if any pipelines fail on the main branch.                                               |
```

## Pipelines running on release branch

The following CI jobs generate release artifacts and verify the functional correctness of Istio Manager.

| Name                                                                                                                        | Description                                                                      |
|-----------------------------------------------------------------------------------------------------------------------------|----------------------------------------------------------------------------------|

> | [`post-istio-integration`](https://github.com/kyma-project/test-infra/blob/main/prow/jobs/istio/istio-integration.yaml#L87) | Executes main integration test suite for Istio Module on k3d cluster.            |

Reviewer: ```suggestion
| [`post-istio-integration`](https://github.com/kyma-project/test-infra/blob/main/prow/jobs/istio/istio-integration.yaml#L87) | Executes the main integration test suite for the Istio module on a k3d cluster.            |
```

> | [`rel-istio-build`](https://github.com/kyma-project/test-infra/blob/main/prow/jobs/istio/istio-manager.yaml#L208)           | Builds Istio operator image on the release and pushes it into the prod registry. |

Reviewer: ```suggestion
| [`rel-istio-build`](https://github.com/kyma-project/test-infra/blob/main/prow/jobs/istio/istio-manager.yaml#L208)           | Builds Istio Operator's image on the release and pushes it to the `prod` registry. |
```

> | [`rel-istio-module-build`](https://github.com/kyma-project/test-infra/blob/main/prow/jobs/istio/istio-manager.yaml#L263)    | Build Istio module template on the release and pushes it into the prod registry. |

Reviewer: ```suggestion
| [`rel-istio-module-build`](https://github.com/kyma-project/test-infra/blob/main/prow/jobs/istio/istio-manager.yaml#L263)    | Builds Istio ModuleTemplate on the release and pushes it to the `prod` registry. |
```

## Pipelines running on a schedule

> The following CI jobs runs on a given schedule.

Reviewer: ```suggestion
The following CI jobs are scheduled to run at specific times.
```

| Name                                                                                                                                 | Description                                                                         |
|--------------------------------------------------------------------------------------------------------------------------------------|-------------------------------------------------------------------------------------|

> | [`nightly-istio-performance-test-gcp`](https://github.com/kyma-project/test-infra/blob/main/prow/jobs/istio/istio-manager.yaml#L365) | Runs Istio Manager performance tests every day at 12:00 AM on Gardener GCP cluster. |

Reviewer: ```suggestion
| [`nightly-istio-performance-test-gcp`](https://github.com/kyma-project/test-infra/blob/main/prow/jobs/istio/istio-manager.yaml#L365) | Runs Istio Manager's performance tests every day at 12:00 AM on a Gardener GCP cluster. |
```

# CI/CD

> The Istio module contains several pipelines that run on pull requests, on the main branch, on the release branch, and periodically. They perform various tasks such as running a linter, executing unit tests with code coverage, generating artifacts for development and production purposes, and running integration tests. The integration tests ensure the functional correctness of Istio Manager and verify if upgrading to the new version of Istio Manager is possible.

Reviewer: ```suggestion
The Istio module contains several pipelines that run on pull requests, on the main branch, on the release branch, and periodically. They perform various tasks such as running a linter, executing unit tests with code coverage, generating artifacts for development and production purposes, and running integration tests. The integration tests ensure the functional correctness of Istio Operator and verify if upgrading to the new version of Istio Operator is possible.
```
